### PR TITLE
ELPP-1759 Article ingestion via Drush.

### PR DESCRIPTION
**Pull request by nlisgo**
Added setting for Lax to settings.php.

ELPP-1759 Update to get all pages of articles.

ELPP-1759 Removed references to ingest and replaced them with import-all.

ELPP-1759 Work in progress.

ELPP-1759 Work in progress.

ELPP-1759 Fixed mass importation.

### CODE REVIEW

**Comment by nlisgo** [comment]
The article import appeared to complete ok but it seemed to go beyond the number of articles that it queried:

```
Processed article 00242 (3328 of 3283)
Processed 3283 articles in 43 minutes 50 seconds.
```

File: src/modules/jcms_article/src/FetchArticle.php [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+   */
+  public function getAllArticleIds() {
+    $ids = [];
+    $articles = $this->getAllArticles();
+    if ($articles) {
+      foreach ($articles as $article) {
+        $ids[] = $article->getId();
+      }
+    }
+    return array_values(array_unique($ids));
```
Comment (nlisgo) [inline src/modules/jcms_article/src/FetchArticle.php]: This fixes the issue with the progress counter:
```
Processed article 00242 (3283 of 3283)
Processed 3283 articles in 42 minutes 54 seconds.
```

**Comment by nlisgo** [comment]
No preview articles have been imported. There were 27 article snippets that were invalid and 45 duplicate article ids.

File: src/modules/jcms_article/src/FetchArticle.php [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      $page = 1;
+      while (!$stop) {
+        $response = $this->client->get($endpoint, ['auth' => Settings::get($auth_key), 'http_errors' => FALSE, 'query' => ['page' => $page]]);
+        if ($response instanceof ResponseInterface) {
+          $body = $response->getBody()->getContents();
+          $json = json_decode($body, TRUE);
+          if (isset($json['items']) && !empty($json['items'])) {
+            foreach ($json['items'] as $data) {
+              if (isset($data['id'])) {
+                $articles[$data['id']] = new Article($data['id'], json_encode($data));
```
Comment (nlisgo) [inline src/modules/jcms_article/src/FetchArticle.php]: This ensures that we only store an article once in the array and makes it easier to check how many articles we can limit the import to, if a limit is set.

File: src/modules/jcms_article/src/FetchArticle.php [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+   */
+  public function getAllArticleIds() {
+    $ids = [];
+    $articles = $this->getAllArticles();
+    if ($articles) {
+      foreach ($articles as $article) {
+        $ids[] = $article->getId();
+      }
+    }
+    return array_values($ids);
```
Comment (nlisgo) [inline src/modules/jcms_article/src/FetchArticle.php]: This strips out the keys which are article ids.

File: src/modules/jcms_article/jcms_article.module [hunk 4]
```diff
@@ -34,6 +34,6 @@ function jcms_article_node_presave(EntityInterface $entity) {
   $node_presave = \Drupal::service('jcms_article.hooks.node_presave');
   $node_presave->addJsonFields($entity);
   $node_presave->setPublishedStatus($entity);
-  $node_presave->setPublishedDate($entity);
+  $node_presave->setStatusDate($entity);
```
Comment (nlisgo) [inline src/modules/jcms_article/jcms_article.module]: The status date needs to occupy the created field for more appropriate ordering of articles in lists.

File: src/modules/jcms_article/src/Entity/ArticleVersions.php [hunk 5]
```diff
@@ -39,7 +39,7 @@
    * @param string $json
    * @param int $action
    */
-  public function __construct(string $id, string $json, int $action = self::WRITE) {
+  public function __construct(string $id, string $json = '', int $action = self::WRITE) {
```
Comment (nlisgo) [inline src/modules/jcms_article/src/Entity/ArticleVersions.php]: This allows an article node to be created by just supplying an article id which is utilised on migrate.

File: src/modules/jcms_migrate/src/Plugin/migrate/source/JCMSCollectionNode.php [hunk 6]
```diff
@@ -32,8 +32,8 @@ public function query() {
     $query->addField('impact', 'field_elife_c_text_value', 'summary');
     $query->addField('fm', 'uri', 'photo_uri');
     $query->addExpression("GROUP_CONCAT(DISTINCT curators.field_elife_c_curators_target_id ORDER BY curators.delta ASC SEPARATOR '|')", 'curators');
-    $query->addExpression("GROUP_CONCAT(IF(an.type = 'elife_news_article', CONCAT('blog_article|', an.nid), IF(an.type = 'elife_article_reference', CONCAT('article|', SUBSTRING_INDEX(an.title, ':', 1)), an.type)) ORDER BY articles.delta SEPARATOR '||')", "content");
-    $query->addExpression("GROUP_CONCAT(SUBSTRING_INDEX(rn.title, ':', 1) ORDER BY related.delta ASC)", 'related');
+    $query->addExpression("GROUP_CONCAT(DISTINCT IF(an.type = 'elife_news_article', CONCAT('blog_article|', an.nid), IF(an.type = 'elife_article_reference', CONCAT('article|', SUBSTRING_INDEX(an.title, ':', 1)), an.type)) ORDER BY articles.delta SEPARATOR '||')", "content");
+    $query->addExpression("GROUP_CONCAT(DISTINCT SUBSTRING_INDEX(rn.title, ':', 1) ORDER BY related.delta ASC)", 'related');
```
Comment (nlisgo) [inline src/modules/jcms_migrate/src/Plugin/migrate/source/JCMSCollectionNode.php]: Addresses issue with import of some collections.

File: src/modules/jcms_notifications/drush/jcms_notifications.drush.inc [hunk 7]
```diff
@@ -9,23 +9,75 @@
...
       'drush article-import' => 'Import articles from Lax and return a message when finished.',
-    ),
-  );
+    ],
+  ];
+  $items['article-import-all'] = [
+    'description' => 'Imports all articles from Lax.',
+    'options' => [
+      'limit' => 'Limit on the number of items to process in each import.',
+      'skip-updates' => 'Do not attempt to update articles that exist already.',
```
Comment (nlisgo) [inline src/modules/jcms_notifications/drush/jcms_notifications.drush.inc]: Added some useful options.

File: src/modules/jcms_rest/src/Plugin/rest/resource/AbstractRestResourceBase.php [hunk 8]
```diff
@@ -582,17 +478,22 @@ public function getEntityQueueItem(EntityInterface $node, FieldItemListInterface
     $item_values = [
       'title' => $node->getTitle(),
       'image' => $this->processFieldImage($node->get('field_image'), TRUE, 'banner', TRUE),
-      'item' => [],
     ];
 
     if ($related->getType() == 'article') {
-      $item_values['item'] = $this->getArticleSnippet($related);
+      if ($article = $this->getArticleSnippet($related)) {
+        $item_values['item'] = $article;
```
Comment (nlisgo) [inline src/modules/jcms_rest/src/Plugin/rest/resource/AbstractRestResourceBase.php]: This and similar changes help us cope with the fact that some article json is still invalid.

Comment (giorgiosironi) [inline src/modules/jcms_rest/src/Plugin/rest/resource/AbstractRestResourceBase.php]: With latest updates that need to be rolled out to prod) are are at only 1 invalid article in the corpus, getting there.

**Pull request merged**